pkg/controller/servicebindingrequest/servicecontext.go

```diff
@@ -0,0 +1,146 @@
+package servicebindingrequest
+
+import (
+	"github.com/imdario/mergo"
+	"github.com/redhat-developer/service-binding-operator/pkg/controller/servicebindingrequest/annotations"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+
+	v1alpha1 "github.com/redhat-developer/service-binding-operator/pkg/apis/apps/v1alpha1"
+)
+
+// ServiceContext contains information related to a service.
+type ServiceContext struct {
+	// Object is the resource being used as reference.
+	Object *unstructured.Unstructured
+	// EnvVars contains the service's contributed environment variables.
+	EnvVars map[string]interface{}
+	// VolumeKeys contains the keys that should be mounted as volume from the binding secret.
+	VolumeKeys []string
+	// EnvVarPrefix indicates the prefix to use in environment variables.
+	EnvVarPrefix *string
+}
+
+// ServiceContextList is a list of ServiceContext values.
+type ServiceContextList []*ServiceContext
+
+// GetObjects returns a slice of service unstructured objects contained in the collection.
+func (sc ServiceContextList) GetObjects() []*unstructured.Unstructured {
+	var crs []*unstructured.Unstructured
+	for _, s := range sc {
+		crs = append(crs, s.Object)
+	}
+	return crs
+}
+
+// buildServiceContexts return a collection of ServiceContext values from the given service
+// selectors.
+//
+// TODO(isuttonl): implement tests
+func buildServiceContexts(
+	client dynamic.Interface,
+	ns string,
+	selectors []v1alpha1.BackingServiceSelector,
+	envVarPrefix *string,
+) (ServiceContextList, error) {
+	serviceCtxs := make([]*ServiceContext, 0)
+	for _, s := range selectors {
+		if s.Namespace == nil {
+			s.Namespace = &ns
+		}
+
+		gvk := schema.GroupVersionKind{Kind: s.Kind, Version: s.Version, Group: s.Group}
+
+		serviceCtx, err := buildServiceContext(
+			client, *s.Namespace, gvk, s.ResourceRef, envVarPrefix)
+		if err != nil {
+			return nil, err
+		}
+		serviceCtxs = append(serviceCtxs, serviceCtx)
+	}
+
+	return serviceCtxs, nil
+}
+
+func buildServiceContext(
```

Comment:
very large function, hard to follow, can we split/extract logical units into separate functions?

Reply:
Can you comment if we can do something about it?

Reply:
It can be simplified, yes.